File: services/iam-service/src/iam/iam.service.ts
import { Injectable, Logger, NotFoundException, BadRequestException } from '@nestjs/common';
import { randomUUID } from 'node:crypto';
import { DatabaseService } from '../database/database.service';
import { RequestContextService } from '../observability/request-context.service';
import {
  AssignRolesDto,
  CreateUserDto,
  ListUsersQueryDto,
  UpdateUserDto,
} from './dto/iam.dto';

export interface UserRow {
  id: string;
  email: string;
  display_name: string;
  tenant_id: string;
  active: boolean;
  created_at: Date;
  updated_at: Date;
}

export interface RoleRow {
  id: string;
  name: string;
}

@Injectable()
export class IamService {
  private readonly logger = new Logger(IamService.name);

  constructor(
    private readonly db: DatabaseService,
    private readonly requestContext: RequestContextService,
  ) {}

  async createUser(dto: CreateUserDto) {
    const id = randomUUID();
    const result = await this.db.query<UserRow>(
      `INSERT INTO iam_users (id, email, display_name, tenant_id, active)
       VALUES ($1, $2, $3, $4, TRUE) RETURNING *`,
      [id, dto.email, dto.displayName, dto.tenantId],
    );
    const user = result.rows[0]!;

    if (dto.roleIds && dto.roleIds.length > 0) {
      await this.assignRoleRecords(id, dto.roleIds);
    }

    return this.getUserWithRoles(user.id);
  }

  async listUsers(query: ListUsersQueryDto) {
    const page = query.page ?? 1;
    const pageSize = query.pageSize ?? 20;
    const offset = (page - 1) * pageSize;

    const conditions: string[] = [];
    const params: unknown[] = [];
    let idx = 1;
    if (query.tenantId) {
      conditions.push(`tenant_id = $${idx}`);
      params.push(query.tenantId);
      idx++;
    }
    const where = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';

    const countRes = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM iam_users ${where}`,
      params,
    );
    const total = Number(countRes.rows[0]!.count);

    const data = await this.db.query<UserRow>(
      `SELECT * FROM iam_users ${where} ORDER BY created_at DESC LIMIT $${idx} OFFSET $${idx + 1}`,
      [...params, pageSize, offset],
    );

    const users = await Promise.all(data.rows.map((u) => this.getUserWithRoles(u.id)));
    return { data: users, meta: { page, pageSize, total, totalPages: Math.ceil(total / pageSize) } };
  }

  async getUser(id: string) {
    return this.getUserWithRoles(id);
  }

  async updateUser(id: string, dto: UpdateUserDto) {
    const existing = await this.findUserRow(id);
    const displayName = dto.displayName ?? existing.display_name;
    const email = dto.email ?? existing.email;
    const active = dto.active ?? existing.active;

    const result = await this.db.query<UserRow>(
      `UPDATE iam_users SET display_name = $1, email = $2, active = $3, updated_at = NOW()
       WHERE id = $4 RETURNING *`,
      [displayName, email, active, id],
    );
    return this.getUserWithRoles(result.rows[0]!.id);
  }

  async deactivateUser(id: string) {
    const result = await this.db.query(
      `UPDATE iam_users SET active = FALSE, updated_at = NOW() WHERE id = $1`,
      [id],
    );
    if (result.rowCount === 0) {
      throw new NotFoundException(`User "${id}" not found`);
    }
    this.logger.log({ userId: id }, 'User deactivated');
  }

  async assignRoles(id: string, dto: AssignRolesDto) {
    await this.findUserRow(id);
    await this.assignRoleRecords(id, dto.roleIds);
    return this.getUserWithRoles(id);
  }

  async removeRole(id: string, roleId: string) {
    await this.findUserRow(id);
    const result = await this.db.query(
      `DELETE FROM iam_user_roles WHERE user_id = $1 AND role_id = $2`,
      [id, roleId],
    );
    if (result.rowCount === 0) {
      throw new NotFoundException(`Role "${roleId}" not assigned to user "${id}"`);
    }
    this.logger.log({ userId: id, roleId }, 'Role removed from user');
  }

  async getPermissions(id: string) {
    await this.findUserRow(id);
    const result = await this.db.query<{ permission: string }>(
      `SELECT DISTINCT r.permissions::text as permission
       FROM iam_user_roles ur
       JOIN rbac_roles r ON r.id = ur.role_id
       WHERE ur.user_id = $1`,
      [id],
    );
    const permissions = result.rows
      .map((r) => {
        try {
          return JSON.parse(r.permission) as Array<{
            resource: string;
            action: string;
            conditions?: Record<string, unknown>;
          }>;
        } catch {
          return [];
        }
      })
      .flat();
    return { userId: id, permissions };
  }

  private async assignRoleRecords(userId: string, roleIds: string[]): Promise<void> {
    for (const roleId of roleIds) {
      const roleRes = await this.db.query<RoleRow>(
        `SELECT id, name FROM rbac_roles WHERE id = $1`,
        [roleId],
      );
      if (roleRes.rows.length === 0) {
        throw new BadRequestException(`Role "${roleId}" does not exist`);
      }
      await this.db.query(
        `INSERT INTO iam_user_roles (user_id, role_id) VALUES ($1, $2)
         ON CONFLICT (user_id, role_id) DO NOTHING`,
        [userId, roleId],
      );
    }
  }

  private async findUserRow(id: string): Promise<UserRow> {
    const result = await this.db.query<UserRow>(
      `SELECT * FROM iam_users WHERE id = $1`,
      [id],
    );
    if (result.rows.length === 0) {
      throw new NotFoundException(`User "${id}" not found`);
    }
    return result.rows[0]!;
  }

  private async getUserWithRoles(id: string) {
    const user = await this.findUserRow(id);
    const rolesRes = await this.db.query<RoleRow>(
      `SELECT r.id, r.name FROM iam_user_roles ur
       JOIN rbac_roles r ON r.id = ur.role_id
       WHERE ur.user_id = $1`,
      [id],
    );
    return {
      id: user.id,
      email: user.email,
      displayName: user.display_name,
      tenantId: user.tenant_id,
      active: user.active,
      roles: rolesRes.rows,
      createdAt: user.created_at,
      updatedAt: user.updated_at,
    };
  }
}

File: apps/api/src/api/services/iam_service.py
import uuid
from datetime import datetime, timezone

from fastapi import HTTPException
from sqlalchemy import text


class IamService:
    async def create_user(self, dto, db=None) -> dict:
        user_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc)
        await db.execute(
            text("""
                INSERT INTO iam_users (id, email, display_name, tenant_id, active, created_at, updated_at)
                VALUES (:id, :email, :display_name, :tenant_id, TRUE, :now, :now)
            """),
            {
                "id": user_id,
                "email": dto.email,
                "display_name": dto.display_name,
                "tenant_id": dto.tenant_id,
                "now": now,
            },
        )

        if dto.role_ids:
            for role_id in dto.role_ids:
                await db.execute(
                    text("""
                        INSERT INTO iam_user_roles (user_id, role_id)
                        VALUES (:user_id, :role_id)
                        ON CONFLICT DO NOTHING
                    """),
                    {"user_id": user_id, "role_id": role_id},
                )

        return await self.get_user(user_id, db=db)

    async def list_users(self, page: int, page_size: int, tenant_id: str, db=None) -> tuple[list[dict], int]:
        offset = (page - 1) * page_size

        count_result = await db.execute(
            text("SELECT COUNT(*) FROM iam_users WHERE tenant_id = :tenant_id"),
            {"tenant_id": tenant_id},
        )
        total = count_result.scalar_one() or 0

        rows_result = await db.execute(
            text("""
                SELECT id, email, display_name, tenant_id, active, created_at, updated_at
                FROM iam_users
                WHERE tenant_id = :tenant_id
                ORDER BY created_at DESC
                LIMIT :limit OFFSET :offset
            """),
            {"tenant_id": tenant_id, "limit": page_size, "offset": offset},
        )
        rows = rows_result.fetchall()

        users = []
        for r in rows:
            users.append({
                "id": r[0],
                "email": r[1],
                "display_name": r[2],
                "tenant_id": r[3],
                "active": r[4],
                "created_at": r[5],
                "updated_at": r[6],
                "roles": await self._get_user_roles(r[0], db=db),
            })

        return users, total

    async def get_user(self, user_id: str, db=None) -> dict | None:
        result = await db.execute(
            text("""
                SELECT id, email, display_name, tenant_id, active, created_at, updated_at
                FROM iam_users WHERE id = :id
            """),
            {"id": user_id},
        )
        r = result.fetchone()
        if not r:
            return None

        roles = await self._get_user_roles(user_id, db=db)
        return {
            "id": r[0],
            "email": r[1],
            "display_name": r[2],
            "tenant_id": r[3],
            "active": r[4],
            "created_at": r[5],
            "updated_at": r[6],
            "roles": roles,
        }

    async def update_user(self, user_id: str, dto, db=None) -> dict | None:
        existing = await self.get_user(user_id, db=db)
        if not existing:
            raise HTTPException(status_code=404, detail="User not found")

        sets = []
        params: dict = {"id": user_id}
        if dto.display_name is not None:
            sets.append("display_name = :display_name")
            params["display_name"] = dto.display_name
        if dto.email is not None:
            sets.append("email = :email")
            params["email"] = dto.email
        if dto.active is not None:
            sets.append("active = :active")
            params["active"] = dto.active

        if sets:
            sets.append("updated_at = :now")
            params["now"] = datetime.now(timezone.utc)
            await db.execute(
                text(f"UPDATE iam_users SET {', '.join(sets)} WHERE id = :id"),
                params,
            )

        return await self.get_user(user_id, db=db)

    async def deactivate_user(self, user_id: str, db=None):
        result = await db.execute(
            text("UPDATE iam_users SET active = FALSE, updated_at = :now WHERE id = :id"),
            {"id": user_id, "now": datetime.now(timezone.utc)},
        )
        if result.rowcount == 0:
            raise HTTPException(status_code=404, detail="User not found")

    async def assign_roles(self, user_id: str, role_ids: list[str], db=None):
        existing = await self.get_user(user_id, db=db)
        if not existing:
            raise HTTPException(status_code=404, detail="User not found")

        for role_id in role_ids:
            await db.execute(
                text("""
                    INSERT INTO iam_user_roles (user_id, role_id)
                    VALUES (:user_id, :role_id)
                    ON CONFLICT DO NOTHING
                """),
                {"user_id": user_id, "role_id": role_id},
            )

    async def remove_role(self, user_id: str, role_id: str, db=None):
        result = await db.execute(
            text("DELETE FROM iam_user_roles WHERE user_id = :user_id AND role_id = :role_id"),
            {"user_id": user_id, "role_id": role_id},
        )
        if result.rowcount == 0:
            raise HTTPException(status_code=404, detail="Role assignment not found")

    async def get_permissions(self, user_id: str, db=None) -> list[str]:
        result = await db.execute(
            text("""
                SELECT DISTINCT j.value AS permission
                FROM iam_user_roles ur
                JOIN rbac_roles r ON r.id = ur.role_id
                CROSS JOIN json_each(r.permissions) j
                WHERE ur.user_id = :user_id
            """),
            {"user_id": user_id},
        )
        return [row[0] for row in result.fetchall()]

    async def _get_user_roles(self, user_id: str, db=None) -> list[dict]:
        result = await db.execute(
            text("""
                SELECT r.id, r.name
                FROM iam_user_roles ur
                JOIN rbac_roles r ON r.id = ur.role_id
                WHERE ur.user_id = :user_id
            """),
            {"user_id": user_id},
        )
        return [{"id": row[0], "name": row[1]} for row in result.fetchall()]


iam_service = IamService()
